### PR TITLE
fix: add Message ID validation for IKE CREATE_CHILD_SA and INFORMATION (#995)

### LINKED ISSUE #995: [Bugs] TNGF: IKE Message ID not validated, allowing replay of DELETE/CREATE_CHILD_SA messages
## Bug Description

The TNGF does not validate IKE Message IDs as required by RFC 7296 Section 2.2. The `Dispatch()` function in `NFs/tngf/pkg/ike/dispatcher.go` routes incoming IKE messages to handlers based solely on `ExchangeType` without any Message ID verification. Individual handlers then blindly assign the received Message ID to the IKE SA context without checking it against expected values.

This allows an attacker to **replay captured IKE messages** (which retain valid encryption and integrity) and have them processed as legitimate new requests.

**Vulnerability site 1** — `NFs/tngf/pkg/ike/dispatcher.go` lines 20–60:
```go
func Dispatch(udpConn *net.UDPConn, localAddr, remoteAddr *net.UDPAddr, msg []byte) {
    // ... NATT 4-byte zero check ...
    ikeMessage := new(ike_message.IKEMessage)
    err := ikeMessage.Decode(msg)
    // ...

    switch ikeMessage.ExchangeType {
    case ike_message.IKE_SA_INIT:
        handler.HandleIKESAINIT(udpConn, localAddr, remoteAddr, ikeMessage)
    case ike_message.IKE_AUTH:
        handler.HandleIKEAUTH(udpConn, localAddr, remoteAddr, ikeMessage)
    case ike_message.CREATE_CHILD_SA:
        handler.HandleCREATECHILDSA(udpConn, localAddr, remoteAddr, ikeMessage)
    default:
        ikeLog.Warnf("Unimplemented IKE message type, exchange type: %d", ikeMessage.ExchangeType)
    }
    // ❌ No Message ID validation before dispatching
    // ❌ No INFORMATIONAL handler — DPD and DELETE are silently dropped
}
```

**Vulnerability site 2** — `NFs/tngf/pkg/ike/handler/handler.go` line 478 (IKE_AUTH):
```go
ikeSecurityAssociation.InitiatorMessageID = message.MessageID
// ❌ Directly assigned, should verify MessageID == expected next value
```

**Vulnerability site 3** — `NFs/tngf/pkg/ike/handler/handler.go` line 1108 (CREATE_CHILD_SA):
```go
ikeSecurityAssociation.ResponderMessageID = message.MessageID
// ❌ Directly assigned, no validation
```

**TNGF IKE SA context** — `NFs/tngf/pkg/context/ue.go` lines 146–148:
```go
// Message ID
InitiatorMessageID uint32
ResponderMessageID uint32
// These are stored but never validated against incoming messages
```

Per RFC 7296 Section 2.2:
- Request Message IDs MUST be sequential (each new request increments by 1)
- Response Message IDs MUST match the corresponding request
- Duplicate requests MUST be detected and the previous response retransmitted (not reprocessed)

**Additional finding**: TNGF's `Dispatch()` does **not** handle the `INFORMATIONAL` exchange type (lines 50–58 only handle IKE_SA_INIT, IKE_AUTH, CREATE_CHILD_SA). This means TNGF silently ignores DPD (Dead Peer Detection) and DELETE notifications — a separate issue that compounds the Message ID vulnerability, as stale SAs are never cleaned up.

## To Reproduce

**Source code verification**:

```bash
# No Message ID validation in dispatcher
grep -n "MessageID" NFs/tngf/pkg/ike/dispatcher.go
# No results — field never checked

# Handler blindly assigns MessageID
grep -n "MessageID" NFs/tngf/pkg/ike/handler/handler.go | head -10
# Shows only assignment at lines 478 and 1108, no comparison/validation

# No INFORMATIONAL handler in TNGF dispatcher
grep -n "INFORMATIONAL" NFs/tngf/pkg/ike/dispatcher.go
# No results — exchange type not handled
```

**Structural comparison with N3IWF**:
The TNGF IKE implementation is structurally identical to N3IWF's regarding Message ID handling:
- Same `IKESecurityAssociation` struct with `InitiatorMessageID`/`ResponderMessageID` fields
- Same pattern of direct assignment without validation in handlers
- Same absence of Message ID checks in dispatcher

The TNGF is additionally worse because it lacks the `INFORMATIONAL` handler entirely, meaning DELETE and DPD messages are silently dropped.

## Security Impact

1. **Resource exhaustion via CREATE_CHILD_SA replay (High)** — Replay captured CREATE_CHILD_SA requests to force TNGF to perform expensive DH computations, allocate SPI/XFRM resources, and exhaust kernel XFRM state table.

2. **State machine corruption via IKE_AUTH replay (High)** — Replay old IKE_AUTH messages to reset the IKE SA state machine, overwrite negotiated security parameters, and disrupt ongoing authentication.

3. **Stale SA accumulation (Medium)** — Since TNGF does not handle INFORMATIONAL(DELETE), replaying old CREATE_CHILD_SA messages creates SAs that can never be properly deleted, leading to indefinite resource leak.

4. **Combined with ESP replay-window 0 (Critical)** — Together with the ESP replay protection being disabled (see bug-report-20), an attacker can persistently disrupt any UE's trusted non-3GPP connection.

5. **Enterprise impact** — TNGF handles **trusted** non-3GPP access (e.g., enterprise Wi-Fi). Replay attacks on this path can disrupt all corporate UEs connecting through TNGF, affecting enterprise-wide connectivity.

## Expected Behavior

The TNGF should:

1. Validate IKE Message IDs in `Dispatch()` before routing to handlers, per RFC 7296 Section 2.2:

```go
func Dispatch(udpConn *net.UDPConn, localAddr, remoteAddr *net.UDPAddr, msg []byte) {
    // ... decode ikeMessage ...

    // Look up IKE SA and validate Message ID
    if ikeSA != nil {
        if !ikeMessage.IsResponse() {
            if ikeMessage.MessageID != ikeSA.InitiatorMessageID {
                ikeLog.Warnf("Unexpected request MessageID: got %d, expected %d",
                    ikeMessage.MessageID, ikeSA.InitiatorMessageID)
                return
            }
        } else {
            if ikeMessage.MessageID != ikeSA.ResponderMessageID {
                ikeLog.Warnf("Unexpected response MessageID: got %d, expected %d",
                    ikeMessage.MessageID, ikeSA.ResponderMessageID)
                return
            }
        }
    }
    // ... dispatch to handlers
}
```

2. Add an `INFORMATIONAL` handler to process DELETE and DPD messages properly.

## Environment

- free5GC Version: `v4.2.1`
- OS: `Ubuntu 24.04.2 LTS`
- Kernel version: `6.14.0-37-generic`
- go version: `go1.26.1 linux/amd64`
- Docker image: `free5gc/tngf:v4.2.1`

## Log File

**Source code evidence** :

`NFs/tngf/pkg/ike/dispatcher.go` — Full `Dispatch()` function decodes IKE messages and routes by ExchangeType without any Message ID check. grep confirms zero occurrences of MessageID validation:

```
$ grep -rn "MessageID" NFs/tngf/pkg/ike/dispatcher.go
(no output — field never referenced in dispatcher)

$ grep -n "MessageID" NFs/tngf/pkg/ike/handler/handler.go
478:    ikeSecurityAssociation.InitiatorMessageID = message.MessageID
1108:   ikeSecurityAssociation.ResponderMessageID = message.MessageID
(only blind assignments, no validation)
```

The TNGF IKE Message ID handling is structurally identical to the N3IWF vulnerability, with the additional deficiency of missing INFORMATIONAL exchange support.

### PR DESCRIPTION
**Pull request by warren0813**
Closes https://github.com/free5gc/free5gc/issues/995

Adds minimal IKE Message ID validation to block out-of-order/replayed exchanges.
Wires INFORMATIONAL exchange handling and basic DELETE-path safety.

### CODE REVIEW

File: pkg/ike/dispatcher.go [hunk 1]
```diff
@@ -17,6 +18,69 @@ func init() {
...
+		}
+
+		if _, exists := ikeSecurityAssociation.ThisUE.TemporaryExchangeMsgIDChildSAMapping[ikeMessage.MessageID]; !exists {
+			ikeLog.Warnf(
+				"Unexpected CREATE_CHILD_SA response MessageID: got %d with no pending exchange",
+				ikeMessage.MessageID,
+			)
+			return false
+		}
+
```
[the remark this answers is not in the document]
Comment (warren0813) [reply pkg/ike/dispatcher.go:73]: Fixed as comment above.

File: pkg/ike/dispatcher.go [hunk 2]
```diff
@@ -17,6 +18,69 @@ func init() {
...
+		return true
+	}
+
+	expectedMessageID := ikeSecurityAssociation.PeerRequestMessageID + 1
+	if ikeMessage.MessageID != expectedMessageID {
+		ikeLog.Warnf("Unexpected request MessageID: got %d, expected %d", ikeMessage.MessageID, expectedMessageID)
+		return false
+	}
+	ikeSecurityAssociation.PeerRequestMessageID = ikeMessage.MessageID
+	return true
```
[the remark this answers is not in the document]
Comment (warren0813) [reply pkg/ike/dispatcher.go]: Fixed as comment above.

File: pkg/ike/dispatcher.go [hunk 3]
```diff
@@ -17,6 +18,69 @@ func init() {
...
+	if !ok {
+		// Let handlers process unknown SPI and emit their existing INVALID_IKE_SPI behavior.
+		return true
+	}
+
+	expectedMessageID := ikeSecurityAssociation.PeerRequestMessageID + 1
+	if ikeMessage.MessageID != expectedMessageID {
+		ikeLog.Warnf("Unexpected request MessageID: got %d, expected %d", ikeMessage.MessageID, expectedMessageID)
+		return false
+	}
```
[the remark this answers is not in the document]
Comment (warren0813) [reply pkg/ike/dispatcher.go:95]: Fixed as comment above.

File: pkg/ike/handler/handler.go [hunk 4]
```diff
@@ -1374,6 +1375,74 @@ func HandleCREATECHILDSA(udpConn *net.UDPConn, tngfAddr, ueAddr *net.UDPAddr, me
...
+
+		deletePayload := ikePayload.(*ike_message.Delete)
+		switch deletePayload.ProtocolID {
+		case ike_message.TypeIKE:
+			ikeLog.Info("Received IKE SA DELETE")
+			if ikeSecurityAssociation.ThisUE != nil {
+				ikeSecurityAssociation.ThisUE.Remove()
+			} else {
+				context.TNGFSelf().DeleteIKESecurityAssociation(ikeSecurityAssociation.LocalSPI)
+			}
```
[the remark this answers is not in the document]
Comment (warren0813) [reply pkg/ike/handler/handler.go:1451]: Fixed as comment above.

**Review by d11nn — CHANGES_REQUESTED** [review]
Hi @warren0813,
I found a few issues that I think should be confirmed or fixed before merge:

1. `PeerRequestMessageID` appears to be advanced too early. At `pkg/ike/dispatcher.go:96`, `ikeSecurityAssociation.PeerRequestMessageID = ikeMessage.MessageID` commits the peer Message ID right after the IKE header is accepted as the next expected ID. However, at that point the request has not yet passed decrypt/authentication/payload handling. If that request is malformed, it can still consume the sequence number and cause a later valid retry to be treated as an old request or a cache-miss duplicate. Could you move the Message ID commit until after the request is successfully validated/processed, or otherwise ensure failed requests do not consume the expected ID?

2. `cachedResponses` does not seem to be tied to the IKE SA / UE lifecycle. Caching responses for duplicate request retransmission makes sense, but the current cleanup looks opportunistic and depends on later packet activity. If the UE disconnects, the IKE SA is deleted, or the connection becomes idle, stale responses may remain in the `sync.Map`. Could you add explicit cleanup on IKE SA / UE teardown, or a TTL/background cleanup mechanism?

3. `TemporaryExchangeMsgIDChildSAMapping` has no timeout. At `pkg/context/ue.go:300`, `ue.TemporaryExchangeMsgIDChildSAMapping[msgID] = childSA` stores a pending Child SA by Message ID, but I do not see timeout or failure cleanup. If the CREATE_CHILD_SA exchange never completes, this pending entry can remain indefinitely and may affect later Message ID or Child SA state handling. Could you add timeout/failure cleanup for this mapping?

4. Could you add tests covering how TNGF distinguishes new requests, duplicate retransmissions, malformed requests, and expired/stale requests? In particular, it would be useful to test that malformed requests do not consume `PeerRequestMessageID`, and duplicate requests only replay responses that were successfully produced before.

File: pkg/ike/handler/handler.go [hunk 5]
```diff
@@ -1401,6 +1377,95 @@ func HandleCREATECHILDSA(udpConn *net.UDPConn, tngfAddr, ueAddr *net.UDPAddr, me
...
+
+			for offset := 0; offset+4 <= len(deletePayload.SPIs); offset += 4 {
+				inboundSPI := binary.BigEndian.Uint32(deletePayload.SPIs[offset : offset+4])
+				context.TNGFSelf().ChildSA.Delete(inboundSPI)
+				if ikeSecurityAssociation.ThisUE != nil {
+					delete(ikeSecurityAssociation.ThisUE.TNGFChildSecurityAssociation, inboundSPI)
+				}
+			}
+		}
```
Comment (d11nn) [inline pkg/ike/handler/handler.go]: `TNGFChildSecurityAssociation` is a plain` map[uint32]*ChildSecurityAssociation` on `TNGFUe` without mutex protection.

Deleting from this map while other goroutines are reading/writing it (e.g. during packet processing or child SA
lookup) can trigger a Go runtime `fatal error: concurrent map read and map write`, which cannot be caught by `recover()` and will crash the process immediately.

**Suggestion**: Add a mutex to `TNGFUe` (similar to how you added `TemporaryExchangeMsgIDChildSAMu`) or provide thread-safe helper methods (e.g., `DeleteChildSA(spi uint32)`) to protect this map.

File: pkg/ike/handler/send.go [hunk 6]
```diff
@@ -2,25 +2,71 @@ package handler
...
+func init() {
+	go func() {
+		ticker := time.NewTicker(cachedResponseCleanupInterval)
+		defer ticker.Stop()
+
+		for now := range ticker.C {
+			cleanupExpiredCachedResponses(now)
+		}
+	}()
+}
```
Comment (d11nn) [inline pkg/ike/handler/send.go]: Because this background goroutine runs independently outside of `Dispatch()`, any unhandled panic inside it (such as from the unsafe type assertion `response := value.(cachedResponse)` in `cleanupExpiredCachedResponses`) will crash the entire TNGF process without being recovered.

**Suggestion**:
1. Add `defer func() { if r := recover(); r != nil { ... } }()` inside the background goroutine.
2. Use safe comma-ok type assertions in `cleanupExpiredCachedResponses`

**Review by d11nn — CHANGES_REQUESTED** [review]

**Comment by d11nn** [comment]
@Alonza0314 LGTM

**Review by d11nn — APPROVED** [review]

**Comment by Alonza0314** [comment]
@warren0813 Please help to rebase or merge from main branch and re-run the IT again!

**Review by Alonza0314 — APPROVED** [review]
